main.py

```diff
@@ -534,52 +534,80 @@ async def list_result_directory() -> list[int]:
+            await websocket.send_json(datasets)
+    except websockets.exceptions.ConnectionClosedError:
+        logger.info("The connection for sending the dataset list is closed.")
+    except websockets.exceptions.WebSocketException:
+        logger.exception("Failed to send the dataset list.")
 
-@app.get("/dataset/master/modification/")
-async def get_dataset_modification(
-    key: str, timestamp: float, timeout: Optional[float],
-) -> tuple[float, tuple[dset.Modification, ...]]:
-    """Returns the dataset modifications since the given timestamp.
-    
-    See dataset.DatasetTracker.since() for details.
+
+@app.websocket("/dataset/master/modification/")
+async def get_dataset_modification(websocket: WebSocket):
+    """Sends the specific dataset modification whenever it is modified.
+
+    After accepted, it receives the target dataset name.
+    Then, it sends the current dataset, parameters, and units immediately.
+    Finally, it sends the dataset modificiation at least a second apart, every time it is modified.
+
+    For details about dataset modificiation, see dataset.DatasetTracker.since().
 
     Args:
-        key: The key of the target dataset.
-        timestamp: The timestamp of the last update.
-        timeout: The timeout in seconds for awaiting new modifications.
-          None for no timeout (wait until done), and 0 or negative for non-blocking.
-          The actual execution time might exceed the timeout because of the
-          since() calls are not timed. Although the timeout becomes less precise,
-          it can guarantee a valid return value for even short timeouts.
+        websocket: The web socket object.
     """
-    latest, modifications = dataset_tracker.since(key, timestamp)
-    if modifications or latest < 0 or (timeout is not None and timeout <= 0):
-        return latest, modifications
+    await websocket.accept()
     try:
-        await asyncio.wait_for(dataset_tracker.modified[key].wait(), timeout)
-    except asyncio.TimeoutError:
-        return latest, modifications
-    return dataset_tracker.since(key, timestamp)
+        name = await websocket.receive_json()
+        timestamp, dataset = dataset_tracker.get(name)
+        await websocket.send_json(dataset)
+        _, parameters = dataset_tracker.get(f"{name}.parameters")
+        await websocket.send_json(parameters)
+        _, units = dataset_tracker.get(f"{name}.units")
+        await websocket.send_json(units)
+        while True:
+            await asyncio.wait_for(dataset_tracker.modified[name].wait(), None)
+            latest, modifications = dataset_tracker.since(name, timestamp)
+            if latest < 0:  # dataset is overwritten or removed
+                await websocket.send_json(None)
+                break
+            if latest > timestamp + 1:  # changes within 1 second are buffered
+                timestamp = latest
+                await websocket.send_json(modifications)
```

Comment:
This loop runs too busy even though it does not send anything within 1 second.
How about inserting a [`sleep()`](https://docs.python.org/3/library/asyncio-task.html#asyncio.sleep) after sending modifications, instead of checking the timestamp?

Reply:
Why couldn't I think of this simple way?😮 I will apply your idea.